include/bluetooth/audio/csis.h

```diff
@@ -292,14 +292,15 @@ void bt_csis_client_register_cb(struct bt_csis_client_cb *cb);
 
 /**
  * @brief Read the lock value of a specific device and instance.
- *
- * @param conn      Pointer to the connection to the device.
- * @param inst_idx  Index of the CSIS index of the peer device (as it may have
- *                  multiple CSIS instances).
+
+ * @param members   Member to get the lock value from
+ * @param set       Pointer to the specific set of the @p member to get the
+ *                  lock value from.
  *
  * @return Return 0 on success, or an errno value on error.
  */
-int bt_csis_client_lock_get(struct bt_conn *conn, uint8_t inst_idx);
+int bt_csis_client_lock_get(struct bt_csis_client_set_member *member,
+			    const struct bt_csis_client_set *set);
```

Comment:
Two minor comments or rather questions from me:
1. First of all, do the API user needs to read the lock state by itself? Is this function needed to be exposed in public API? Shouldn't the user call to acquire and release the lock if needed only?
2. Do you think the function name suffix *_get* reminds *_acquire*? But it's not what this function does actually. Maybe you could consider changing name to *_get_state*.

Reply:
1. It's a good question, and the answer is "maybe not". A user may want to ensure that all set members are unlocked before attempting to lock the set (as recovering from a failed set lock procedure is more expensive than reading the lock state on all devices first. Furthermore, we will need this to pass qualification tests iirc. 

2. I had never considered that `get` may be misunderstood in this case, but I definitely see your point. `get_state` is much better.

Reply:
Ad. 1. But isn't the CSIS implementation job to track the member lock state? I guess it should do it, because it subscribes to Set Member Lock ATT Notification. So the CSIS implementation knows a priori whether member is locked and whether it's locked by us before user calls API bt_csis_client_lock/bt_csis_client_release.
As API user I would expect the CSIS to do acquire/release when requested and/or return success or error on failure.
If the logic will be done in this service, one would not need to handle it in upper layers e.g. CAP.

Regarding qualification test, if there is a test case that requires ATT Read Request to Lock State we could pass it anyway if the stack will read the Lock State right after connection. It has to be done anyway IMO knowing the state, implementation decides whether to request or not the Lock if it's already granted to someone else.

Ad. 2 Thanks ^^

Reply:
Assuming that we read the lock state value when connect to a non-bonded device, then we would still need to do the discovery first. It isn't something that we have considered much before, but we don't usually do discovery on connection established autonomously in Zephyr, which would be required to read the lock state. Due to that dependency, we can't auto-read the state (although we could read it as part of discovery procedure).
For bonded devices, we should expect a notification on reconnection. 
We will also need to assume that notifications are not lost on air (or otherwise), if we don't want to expose an explicitly read procedures. 

If we need to implement the functionality in the first place, it could also be argued that it doesn't hurt to expose this as an API.  

So the way I see it, we have the options of
1) Auto discover and auto read lock state on connection
2) Read lock state during discovery procedure
3) Expose the get_lock_state function

Where 1) is a major change, 3) is what we have and 2) can fairly easily be implemented.

Reply:
I know that Zephyr don't discover ATT attributes after ACL is established autonomously.
I was thinking rather that the Lock State read could be part if bt_csis_client_discover (sorry I was unclear).

Reply:
Given that there isn't a defined procedure for it in the spec, I think it would be fine to keep in the discovery. 

Will update this PR with a new commit that changes that.

Reply:
I will have to read the whole PR to get the full picture, but here are some inputs on this point.
- Doing discovery autonomously and directly after connection could be a pretty large load on the system in the cases where there are many services (as there may be in le-audio).  I believe some applications would rather prefer to be in control of this.
- For the lock characteristic, notifications are mandatory, but for the other characteristics of CSIS, notifications are not mandatory.  So, leaving out the read (API) for the lock would make it different than the other characteristics.  (Which may be OK, or may be seen as lack of consistency.)
- Even if the notifications are mandatory, notifications are not reliable (over ATT).  So the state kept may be wrong.  (Not an issue over EATT, but we have had as a policy to not rely upon, or mandate, EATT.
- Since the "get" was mentioned: "set" is also open for interpretation - does lock_set() lock the set, or does it set the lock? (Not really a problem in that case, as locking the set implies setting the lock ...)

Reply:
> * Even if the notifications are mandatory, notifications are not reliable (over ATT).  So the state kept may be wrong.
could you point me to the 

Could you please point out the place in the Zephyr Bluetooth code, where Notification packet could be dropped, just because this is notification or because of the lack of the buffers?
I have feeling this is not a case for the Zephyr. Is so then saying "notification as not reliable" is incorrect.

Reply:
> * Doing discovery autonomously and directly after connection could be a pretty large load on the system in the cases where there are many services (as there may be in le-audio).  I believe some applications would rather prefer to be in control of this.

@asbjornsabo What would be policy for the discovery start for such an applications if not the Connection Establishment time? I'm curious :)

Reply:
> Could you please point out the place in the Zephyr Bluetooth code, where Notification packet could be dropped, just because this is notification or because of the lack of the buffers?

We don't do that in Zephyr AFAIK, but that doesn't mean other implementations won't. 

> What would be policy for the discovery start for such an applications if not the Connection Establishment time? I'm curious :)

Application specific I suppose, of if requirement from other profiles :) 

LE Audio is a bit unique in this regard, where both the central and peripheral need to do discovery, as otherwise that has mostly been the central, so it would not make sense to for a Zephyr device acting as a peripheral to always do a discovery.

Reply:
> We don't do that in Zephyr AFAIK, but that doesn't mean other implementations won't.
> 
> >

Agree, then we are good. As a Client we can rely on the notifications because those are reliable on Zephyr.
And as a Server we will follow what other Client expects as to do and leave that issue to solve to other implementations.

Reply:
@rymanluk Right, I think I finally understood what you mean. 

Since the dropping of notifications is a GATT client only issue, then the server won't care about that. 
The Zephyr client will never drop notifications due to buffer issues (perhaps it should?), but Zephyr does not need this procedure to function properly as a set coordinator. 

The point is moot though, as the procedure is part of the CAP test spec (details not public), so we will need to perform this procedure to qualify CAP, unless we change the CAP test spec. 

I suggest that if this is the only blocker for this PR, that we get it merged and then if we can avoid the procedure all-together due to the specification changes, then we can remove it for optimization. Does that sound OK to you?

Reply:
@Thalley  That is fine, however I would not worry about CAP Test spec - because the only thing we need to do is read Lock, and this we need to do anyway after first connection.

Reply:
@rymanluk But if the test have the sequence:

1) CSIP discover (discovery + read lock)
2) CAP discover
3) CSIP Read Ordered Access
4) Do some other procedure 
(which is the case in some CAP tests)

Then we would otherwise need to do csip_discover() for step 3 again to do the read of the lock. So we could probably hack our way into passing the test, but we'd need to do discovery do perform a lock read if we don't have this procedure implemented.